builder/simplify.cc

```diff
@@ -373,8 +383,8 @@ class simplifier : public node_mutator {
 
     // First, prune any dead lets, and any lets that are just vars
     for (auto it = lets.begin(); it != lets.end();) {
-      int refs = *references[it->first];
-      if (refs == 0 || is_variable(it->second, it->first)) {
+      auto& ref_info = references[it->first];
+      if (!ref_info || ref_info->counter_ == 0 || is_variable(it->second, it->first)) {
```

Comment:
The existing line of code here is very buggy, it deletes lets if they are a single variable without substituting. This should look more like the `let` visitor now that we don't need to be careful about substitution.